cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_custom_slurm_settings_include_file_generator.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file.
+# This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+import argparse
+import logging
+from os import makedirs, path
+from typing import Dict, List
+
+import yaml
+
+INCLUDE_FILE_HEADER = "# This file is automatically generated by pcluster"
+INCLUDE_TARGET_MAPPING = {
+    "slurm": "CustomSlurmSettings",
+}
+log = logging.getLogger()
+
+
+def _load_cluster_config(input_file_path):
+    """
+    Load cluster configuration to get info used to render templates.
+
+    :param input_file_path: path to the cluster configuration YAML file
+    :return: dictionary containing the cluster configuration structure
+    """
+    with open(input_file_path, encoding="utf-8") as input_file:
+        return yaml.load(input_file, Loader=yaml.SafeLoader)
+
+
+def _setup_logger():
+    logging.basicConfig(
+        level=logging.INFO, format="%(asctime)s - [%(name)s:%(funcName)s] - %(levelname)s - %(message)s"
+    )
+
+
+def _generate_simple_parameter(param: Dict) -> str:
+    """
+    Generate string to represent a simple parameter in the Slurm configuration.
+
+    :param param: dictionary containing the parameter structure.
+    :return: string to be included in the Slurm configuration file.
+    """
+    key = list(param.keys())[0]
+    value = list(param.values())[0]
+    return f"{key}={value}"
+
+
+def _generate_complex_parameter(param: Dict) -> str:
+    """
+    Generate string to represent a complex parameter in the Slurm configuration.
+
+    :param param: dictionary containing the parameter structure.
+    :return: string to be included in the Slurm configuration file.
+    """
+    # Some complex parameters require a specific subparameter to be defined first in the line
+    # (e.g. the NodeName parameter in the node definition)
+    for key, value in param.items():
+        # Here we assume that the users do not pass more than one of these parameters in the same dictionary.
+        if key.lower() in ["nodename", "partitionname", "nodeset", "downnodes"]:
```

Comment:
> We should check that no more than one of these keys is present or fail.

I'm not sure I agree with that.

I mean, I see the problem, but that would be a sort of validation performed at cookbook time: we should either do it in the CLI at validator time, or not do it at all since a main assumption in the feature is that we are not going to validate the Slurm config coming from the user, other than checking against the deny-list. Let's discuss about this.

> Also, unless you want to make this function more general and handle both cases you should also check that at least one is present.

Yes, I'm going to handle both cases with the same method. Plus I wonder if there are other configuration parameters that may be composed of space-separated key=value pairs that are not covered by that list...

Reply:
OK, we will put a warning if more than one of the special parameters in the list are provided.

Reply:
Addressed.